File: mobile/ios/OnlyCoffee/Models/Loyalty.swift
import Foundation

// MARK: - User Tier Enum
enum UserTier: String, Codable {
    case bronze = "bronze"
    case silver = "silver"
    case gold = "gold"
    case platinum = "platinum"
    case black = "black"

    var displayName: String {
        rawValue.capitalized
    }

    var color: String {
        switch self {
        case .bronze: return "bronze"
        case .silver: return "gray"
        case .gold: return "yellow"
        case .platinum: return "blue"
        case .black: return "black"
        }
    }
}

// MARK: - User Streak Model
struct UserStreak: Codable, Identifiable {
    let id: String
    let userId: String
    let consecutiveDays: Int
    let currentStreak: Int
    let longestStreak: Int
    let lastVisitDate: Date?
    let firstQualifyingPurchaseDate: Date?
    let streakStartDate: Date?
    let monthlyPoints: Int
    let tierXP: Int
    let monthlyVisits: Int
    let annualSpend: Double
    let sevenDayStreakCount: Int
    let lastMonthlyReset: Date?
    let createdAt: Date
    let updatedAt: Date
}

// MARK: - Streak Visit Model
struct StreakVisit: Codable, Identifiable {
    let id: String
    let userId: String
    let orderId: String
    let visitDate: Date
    let orderAmount: Double
    let isMorningRush: Bool
    let pointsEarned: Int
    let xpEarned: Int
    let basePoints: Int
    let baseXP: Int
    let multiplier: Int
    let streakDayAtVisit: Int
    let createdAt: Date

    var multiplierText: String {
        isMorningRush ? "2x MORNING RUSH" : "1x"
    }
}

// MARK: - Streak Saver Token Status
enum TokenStatus: String, Codable {
    case available = "available"
    case used = "used"
    case expired = "expired"
}

// MARK: - Streak Saver Token Model
struct StreakSaverToken: Codable, Identifiable {
    let id: String
    let userId: String
    let status: TokenStatus
    let grantedAt: Date
    let usedAt: Date?
    let appliedToDate: Date?
    let expiresAt: Date?
    let metadata: [String: AnyCodable]?
    let createdAt: Date

    var isAvailable: Bool {
        status == .available && !isExpired
    }

    var isExpired: Bool {
        guard let expiresAt = expiresAt else { return false }
        return Date() >= expiresAt
    }

    var expiryText: String {
        guard let expiresAt = expiresAt else { return "No expiry" }

        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d"
        formatter.timeZone = TimeZone(identifier: "America/Chicago")
        return "Expires \(formatter.string(from: expiresAt))"
    }
}

// MARK: - Streak Reward Model
struct StreakReward: Codable, Identifiable {
    let id: String
    let streakDay: Int
    let rewardType: String
    let couponType: CouponType
    let label: String
    let description: String?
    let valueCents: Int?
    let maxValueCents: Int?
    let expiryDays: Int
    let channels: String
    let eligibleItems: [String: AnyCodable]?
    let isActive: Bool
    let displayOrder: Int
    let metadata: [String: AnyCodable]?
    let createdAt: Date
    let updatedAt: Date

    var displayValue: String {
        switch couponType {
        case .percentOff:
            return "\(valueCents ?? 0)% OFF"
        case .fixedAmount:
            let amount = Double(valueCents ?? 0) / 100.0
            return String(format: "$%.2f OFF", amount)
        case .freeItem:
            if let maxValue = maxValueCents {
                let max = Double(maxValue) / 100.0
                return String(format: "FREE (up to $%.2f)", max)
            }
            return "FREE"
        case .fixedPrice:
            let price = Double(valueCents ?? 0) / 100.0
            return String(format: "$%.2f", price)
        }
    }
}

// MARK: - Tier Progress Model
struct TierProgress: Codable {
    let currentTier: UserTier
    let nextTier: UserTier?
    let progress: Progress
    let meetsRequirements: Bool

    struct Progress: Codable {
        let monthlyVisits: Metric
        let tierXP: Metric
        let sevenDayStreaks: Metric
        let annualSpend: Metric

        struct Metric: Codable {
            let current: Int
            let required: Int

            var percentage: Double {
                guard required > 0 else { return 0 }
                return min(Double(current) / Double(required), 1.0)
            }
        }
    }
}

// MARK: - Tier Perk Model
struct TierPerk: Codable, Identifiable {
    let id: String
    let tier: UserTier
    let perkType: String
    let perkName: String
    let description: String
    let isActive: Bool
    let displayOrder: Int
    let configuration: [String: AnyCodable]?
    let iconName: String?
    let metadata: [String: AnyCodable]?
    let createdAt: Date
    let updatedAt: Date
}

// MARK: - Next Milestone Model
struct NextMilestone: Codable {
    let nextMilestone: Int?
    let daysUntilMilestone: Int?
    let reward: StreakReward?
}

// MARK: - Anniversary Badge Enum
enum AnniversaryBadge: String, Codable {
    case year1 = "year_1_anniversary"
    case year2 = "year_2_anniversary"
    case year3 = "year_3_anniversary"
    case year5 = "year_5_anniversary"
    case year10 = "year_10_anniversary"
}

// MARK: - Anniversary Reward Model
struct AnniversaryReward: Codable, Identifiable {
    let id: String
    let userId: String
    let anniversaryYear: Int
    let anniversaryDate: Date
    let isGranted: Bool
    let grantedAt: Date?
    let couponId: String?
    let badgeAwarded: AnniversaryBadge?
    let isRedeemed: Bool
    let redeemedAt: Date?
    let customMessage: String?
    let metadata: [String: AnyCodable]?
    let createdAt: Date
}

// MARK: - Next Anniversary Model
struct NextAnniversary: Codable {
    let nextAnniversaryYear: Int?
    let nextAnniversaryDate: Date?
    let daysUntilAnniversary: Int?
    let badge: AnniversaryBadge?
}

// MARK: - Loyalty Dashboard Model
struct LoyaltyDashboard: Codable {
    let streak: StreakInfo
    let tier: TierProgress
    let nextMilestone: NextMilestone
    let streakSaverTokens: Int
    let nextAnniversary: NextAnniversary
    let perks: [TierPerk]

    struct StreakInfo: Codable {
        let consecutiveDays: Int
        let longestStreak: Int
        let monthlyPoints: Int
        let tierXP: Int
        let monthlyVisits: Int
    }
}

// MARK: - API Response Models
struct StreakResponse: Codable {
    let success: Bool
    let data: UserStreak
}

struct VisitsResponse: Codable {
    let success: Bool
    let data: [StreakVisit]
}

struct TokensResponse: Codable {
    let success: Bool
    let data: [StreakSaverToken]
}

struct TokenCountResponse: Codable {
    let success: Bool
    let data: TokenCount

    struct TokenCount: Codable {
        let count: Int
    }
}

struct RewardsResponse: Codable {
    let success: Bool
    let data: [StreakReward]
}

struct NextMilestoneResponse: Codable {
    let success: Bool
    let data: NextMilestone
}

struct TierProgressResponse: Codable {
    let success: Bool
    let data: TierProgress
}

struct TierPerksResponse: Codable {
    let success: Bool
    let data: [TierPerk]
}

struct AnniversariesResponse: Codable {
    let success: Bool
    let data: [AnniversaryReward]
}

struct NextAnniversaryResponse: Codable {
    let success: Bool
    let data: NextAnniversary
}

struct DashboardResponse: Codable {
    let success: Bool
    let data: LoyaltyDashboard
}

struct UseTokenResponse: Codable {
    let success: Bool
    let message: String
    let data: StreakSaverToken?
}

struct VisitedTodayResponse: Codable {
    let success: Bool
    let data: VisitedToday

    struct VisitedToday: Codable {
        let visited: Bool
    }
}

File: android/app/src/main/java/com/onlycoffee/app/data/model/Loyalty.kt
package com.onlycoffee.app.data.model

import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.util.*

// MARK: - User Tier Enum
enum class UserTier {
    @SerializedName("bronze")
    BRONZE,

    @SerializedName("silver")
    SILVER,

    @SerializedName("gold")
    GOLD,

    @SerializedName("platinum")
    PLATINUM,

    @SerializedName("black")
    BLACK;

    val displayName: String
        get() = name.lowercase().replaceFirstChar { it.uppercase() }
}

// MARK: - User Streak Model
@Parcelize
data class UserStreak(
    val id: String,
    @SerializedName("user_id")
    val userId: String,
    @SerializedName("consecutive_days")
    val consecutiveDays: Int,
    @SerializedName("current_streak")
    val currentStreak: Int,
    @SerializedName("longest_streak")
    val longestStreak: Int,
    @SerializedName("last_visit_date")
    val lastVisitDate: Date? = null,
    @SerializedName("first_qualifying_purchase_date")
    val firstQualifyingPurchaseDate: Date? = null,
    @SerializedName("streak_start_date")
    val streakStartDate: Date? = null,
    @SerializedName("monthly_points")
    val monthlyPoints: Int,
    @SerializedName("tier_xp")
    val tierXP: Int,
    @SerializedName("monthly_visits")
    val monthlyVisits: Int,
    @SerializedName("annual_spend")
    val annualSpend: Double,
    @SerializedName("seven_day_streak_count")
    val sevenDayStreakCount: Int,
    @SerializedName("last_monthly_reset")
    val lastMonthlyReset: Date? = null,
    @SerializedName("created_at")
    val createdAt: Date,
    @SerializedName("updated_at")
    val updatedAt: Date
) : Parcelable

// MARK: - Streak Visit Model
@Parcelize
data class StreakVisit(
    val id: String,
    @SerializedName("user_id")
    val userId: String,
    @SerializedName("order_id")
    val orderId: String,
    @SerializedName("visit_date")
    val visitDate: Date,
    @SerializedName("order_amount")
    val orderAmount: Double,
    @SerializedName("is_morning_rush")
    val isMorningRush: Boolean,
    @SerializedName("points_earned")
    val pointsEarned: Int,
    @SerializedName("xp_earned")
    val xpEarned: Int,
    @SerializedName("base_points")
    val basePoints: Int,
    @SerializedName("base_xp")
    val baseXP: Int,
    val multiplier: Int,
    @SerializedName("streak_day_at_visit")
    val streakDayAtVisit: Int,
    @SerializedName("created_at")
    val createdAt: Date
) : Parcelable {
    val multiplierText: String
        get() = if (isMorningRush) "2x MORNING RUSH" else "1x"
}

// MARK: - Streak Saver Token Status
enum class TokenStatus {
    @SerializedName("available")
    AVAILABLE,

    @SerializedName("used")
    USED,

    @SerializedName("expired")
    EXPIRED
}

// MARK: - Streak Saver Token Model
@Parcelize
data class StreakSaverToken(
    val id: String,
    @SerializedName("user_id")
    val userId: String,
    val status: TokenStatus,
    @SerializedName("granted_at")
    val grantedAt: Date,
    @SerializedName("used_at")
    val usedAt: Date? = null,
    @SerializedName("applied_to_date")
    val appliedToDate: Date? = null,
    @SerializedName("expires_at")
    val expiresAt: Date? = null,
    val metadata: Map<String, String>? = null,
    @SerializedName("created_at")
    val createdAt: Date
) : Parcelable {
    val isAvailable: Boolean
        get() = status == TokenStatus.AVAILABLE && !isExpired

    val isExpired: Boolean
        get() = expiresAt?.let { Date().after(it) } ?: false

    val expiryText: String
        get() {
            val expiresAt = this.expiresAt ?: return "No expiry"
            val format = java.text.SimpleDateFormat("MMM d", Locale.getDefault())
            return "Expires ${format.format(expiresAt)}"
        }
}

// MARK: - Streak Reward Model
@Parcelize
data class StreakReward(
    val id: String,
    @SerializedName("streak_day")
    val streakDay: Int,
    @SerializedName("reward_type")
    val rewardType: String,
    @SerializedName("coupon_type")
    val couponType: CouponType,
    val label: String,
    val description: String? = null,
    @SerializedName("value_cents")
    val valueCents: Int? = null,
    @SerializedName("max_value_cents")
    val maxValueCents: Int? = null,
    @SerializedName("expiry_days")
    val expiryDays: Int,
    val channels: String,
    @SerializedName("eligible_items")
    val eligibleItems: Map<String, String>? = null,
    @SerializedName("is_active")
    val isActive: Boolean,
    @SerializedName("display_order")
    val displayOrder: Int,
    val metadata: Map<String, String>? = null,
    @SerializedName("created_at")
    val createdAt: Date,
    @SerializedName("updated_at")
    val updatedAt: Date
) : Parcelable {
    val displayValue: String
        get() = when (couponType) {
            CouponType.PERCENT_OFF -> "${valueCents ?: 0}% OFF"
            CouponType.FIXED_AMOUNT -> {
                val amount = (valueCents ?: 0) / 100.0
                String.format("$%.2f OFF", amount)
            }
            CouponType.FREE_ITEM -> {
                maxValueCents?.let {
                    val max = it / 100.0
                    String.format("FREE (up to $%.2f)", max)
                } ?: "FREE"
            }
            CouponType.FIXED_PRICE -> {
                val price = (valueCents ?: 0) / 100.0
                String.format("$%.2f", price)
            }
        }
}

// MARK: - Tier Progress Model
@Parcelize
data class TierProgress(
    @SerializedName("current_tier")
    val currentTier: UserTier,
    @SerializedName("next_tier")
    val nextTier: UserTier? = null,
    val progress: Progress,
    @SerializedName("meets_requirements")
    val meetsRequirements: Boolean
) : Parcelable {
    @Parcelize
    data class Progress(
        @SerializedName("monthly_visits")
        val monthlyVisits: Metric,
        @SerializedName("tier_xp")
        val tierXP: Metric,
        @SerializedName("seven_day_streaks")
        val sevenDayStreaks: Metric,
        @SerializedName("annual_spend")
        val annualSpend: Metric
    ) : Parcelable {
        @Parcelize
        data class Metric(
            val current: Int,
            val required: Int
        ) : Parcelable {
            val percentage: Float
                get() = if (required > 0) (current.toFloat() / required).coerceAtMost(1.0f) else 0f
        }
    }
}

// MARK: - Tier Perk Model
@Parcelize
data class TierPerk(
    val id: String,
    val tier: UserTier,
    @SerializedName("perk_type")
    val perkType: String,
    @SerializedName("perk_name")
    val perkName: String,
    val description: String,
    @SerializedName("is_active")
    val isActive: Boolean,
    @SerializedName("display_order")
    val displayOrder: Int,
    val configuration: Map<String, String>? = null,
    @SerializedName("icon_name")
    val iconName: String? = null,
    val metadata: Map<String, String>? = null,
    @SerializedName("created_at")
    val createdAt: Date,
    @SerializedName("updated_at")
    val updatedAt: Date
) : Parcelable

// MARK: - Next Milestone Model
@Parcelize
data class NextMilestone(
    @SerializedName("next_milestone")
    val nextMilestone: Int? = null,
    @SerializedName("days_until_milestone")
    val daysUntilMilestone: Int? = null,
    val reward: StreakReward? = null
) : Parcelable

// MARK: - Anniversary Badge Enum
enum class AnniversaryBadge {
    @SerializedName("year_1_anniversary")
    YEAR_1,

    @SerializedName("year_2_anniversary")
    YEAR_2,

    @SerializedName("year_3_anniversary")
    YEAR_3,

    @SerializedName("year_5_anniversary")
    YEAR_5,

    @SerializedName("year_10_anniversary")
    YEAR_10
}

// MARK: - Anniversary Reward Model
@Parcelize
data class AnniversaryReward(
    val id: String,
    @SerializedName("user_id")
    val userId: String,
    @SerializedName("anniversary_year")
    val anniversaryYear: Int,
    @SerializedName("anniversary_date")
    val anniversaryDate: Date,
    @SerializedName("is_granted")
    val isGranted: Boolean,
    @SerializedName("granted_at")
    val grantedAt: Date? = null,
    @SerializedName("coupon_id")
    val couponId: String? = null,
    @SerializedName("badge_awarded")
    val badgeAwarded: AnniversaryBadge? = null,
    @SerializedName("is_redeemed")
    val isRedeemed: Boolean,
    @SerializedName("redeemed_at")
    val redeemedAt: Date? = null,
    @SerializedName("custom_message")
    val customMessage: String? = null,
    val metadata: Map<String, String>? = null,
    @SerializedName("created_at")
    val createdAt: Date
) : Parcelable

// MARK: - Next Anniversary Model
@Parcelize
data class NextAnniversary(
    @SerializedName("next_anniversary_year")
    val nextAnniversaryYear: Int? = null,
    @SerializedName("next_anniversary_date")
    val nextAnniversaryDate: Date? = null,
    @SerializedName("days_until_anniversary")
    val daysUntilAnniversary: Int? = null,
    val badge: AnniversaryBadge? = null
) : Parcelable

// MARK: - Loyalty Dashboard Model
@Parcelize
data class LoyaltyDashboard(
    val streak: StreakInfo,
    val tier: TierProgress,
    @SerializedName("next_milestone")
    val nextMilestone: NextMilestone,
    @SerializedName("streak_saver_tokens")
    val streakSaverTokens: Int,
    @SerializedName("next_anniversary")
    val nextAnniversary: NextAnniversary,
    val perks: List<TierPerk>
) : Parcelable {
    @Parcelize
    data class StreakInfo(
        @SerializedName("consecutive_days")
        val consecutiveDays: Int,
        @SerializedName("longest_streak")
        val longestStreak: Int,
        @SerializedName("monthly_points")
        val monthlyPoints: Int,
        @SerializedName("tier_xp")
        val tierXP: Int,
        @SerializedName("monthly_visits")
        val monthlyVisits: Int
    ) : Parcelable
}

// MARK: - API Response Models
data class StreakResponse(
    val success: Boolean,
    val data: UserStreak
)

data class VisitsResponse(
    val success: Boolean,
    val data: List<StreakVisit>
)

data class TokensResponse(
    val success: Boolean,
    val data: List<StreakSaverToken>
)

data class TokenCountResponse(
    val success: Boolean,
    val data: TokenCount
) {
    data class TokenCount(
        val count: Int
    )
}

data class RewardsResponse(
    val success: Boolean,
    val data: List<StreakReward>
)

data class NextMilestoneResponse(
    val success: Boolean,
    val data: NextMilestone
)

data class TierProgressResponse(
    val success: Boolean,
    val data: TierProgress
)

data class TierPerksResponse(
    val success: Boolean,
    val data: List<TierPerk>
)

data class AnniversariesResponse(
    val success: Boolean,
    val data: List<AnniversaryReward>
)

data class NextAnniversaryResponse(
    val success: Boolean,
    val data: NextAnniversary
)

data class DashboardResponse(
    val success: Boolean,
    val data: LoyaltyDashboard
)

data class UseTokenResponse(
    val success: Boolean,
    val message: String,
    val data: StreakSaverToken?
)

data class VisitedTodayResponse(
    val success: Boolean,
    val data: VisitedToday
) {
    data class VisitedToday(
        val visited: Boolean
    )
}

data class UseTokenRequest(
    @SerializedName("missed_date")
    val missedDate: String
)
